### PR TITLE
chore: Refactor PM Naming Algorithm

### LINKED ISSUE #62: [OTHER] Refactor PM Naming Algorithm
## Summary

Currently, names in the Physical Model (PM) are assigned in different places, leading to implementation complexity. To address this, we need to implement a unified naming algorithm for the PM.

## Context
- A single naming algorithm for the PM is required to streamline the process and eliminate inconsistencies.

### PR DESCRIPTION
**Pull request by aturker-synnada**
## Description

Closes #62 

## What is Changed

The naming algorithm change impacted most of the tests.

-The PM.flatten_dag method has been removed.
- Introduced a new class, FlatModel, whose primary responsibility is to sort primitives in topological order and assign names to all required connections.
- Other test modifications were made accordingly.

## Checklist:

- [x] Tests that cover the code added.
- [x] Corresponding changes documented.
- [x] All tests passed.
- [x] The code linted and styled (pre-commit run --all-files has passed).

### CODE REVIEW

File: tests/scripts/test_compile_keys_consistencies.py [hunk 1]
```diff
@@ -287,4 +287,4 @@ def test_iterable_type_keys():
     for typ in [list, tuple, set, dict]:
         value = dict([("input", None)]) if typ is dict else typ(["input"])
         pm = ml_compile(model, backend, trainable_keys=value)
-        assert pm.data_store.all_static_keys == {"_Linear_0_axes"}
+        assert pm.data_store.all_static_keys == {"axes"}
```
Comment (kberat-synnada) [inline tests/scripts/test_compile_keys_consistencies.py:290]: test_functions (code generation tests) are not updated, how could they pass?

File: tests/scripts/test_flatmodel.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    Linear,
+    Model,
+    Relu,
+    Sigmoid,
+    Softplus,
+    Tanh,
+)
+
+
+def test_flatmodel_with_all_defined():
```
Comment (kberat-synnada) [inline tests/scripts/test_flatmodel.py:30]: Add integration tests

File: tests/json_files/discard_keys_directed_test.json [hunk 3]
```diff
@@ -6,7 +6,7 @@
         "discard_keys": ["output"],
         "results": {
             "output_keys": [],
-            "discard_keys": ["_MatrixMultiply_1_output", "axes", "_Transpose_0_output", "b", "input", "output", "w"]
+            "discard_keys": ["axes", "b", "input", "output", "output_0", "output_1", "w"]
```
Comment (kberat-synnada) [inline tests/json_files/discard_keys_directed_test.json:9]: Naming "output", "output1" etc. to internal connections seems confusing. could we find any other convention to name internal keys?

File: mithril/framework/physical/model.py [hunk 4]
```diff
@@ -1259,3 +1170,351 @@ def evaluate_all(
...
+        self.assigned_names[new_key] = self.assigned_names.pop(old_key)
+
+        if old_key in self.external_mapping.values():
+            self._external_mapping = {
+                key: self.assigned_names[new_key] if value == old_key else value
+                for key, value in self._external_mapping.items()
+            }
+
+    def _name_externals(self):
+        external_keys = list(self.model.conns.input_keys) + list(
```
[the remark this answers is not in the document]
Comment (aturker-synnada) [reply mithril/framework/physical/model.py:1280]: We want to keep order of the keys, due to we want to assign names first to the user named keys.

**Review by norhan-synnada — APPROVED** [review]

**Review by kberat-synnada — APPROVED** [review]

**Pull request merged**